query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,450 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::{
+    Schema,
+};
+
+use mentat_query::{
+    NotJoin,
+    UnifyVars,
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+};
+
+use types::{
+    ColumnConstraint,
+    ComputedTable,
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let projected = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+        let mut template = self.use_as_template(&projected);
+        let mut empty_because: Option<EmptyBecause> = None;
+        
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+            // This is disgusting, but I can't think of anything else right now that works.
+            if template.is_known_empty() {
+                empty_because = template.empty_because;
+                template.empty_because = None;
+            }
+        }
+
+        if template.wheres.is_empty() && empty_because.is_some() {
+            return Ok(());
+        } else {
+            for (v, cols) in (&self.column_bindings).iter().filter(|&(key, _)| projected.contains(&key) ) {
+                match template.column_bindings.entry(v.clone()) {
+                    Entry::Vacant(e) => {
+                        e.insert(cols.clone());
```

Comment:
Here you're collecting every column binding into the `not`. A query like this:

```
[:find ?x ?age ?height
 :where
 [?x :foo/name "John"]
 [?x :foo/age ?age]
 [?x :foo/height ?height]]
```

has `column_bindings` like this:

```
{?x [datoms00.e, datoms01.e, datoms02.e]}
```

and produces a query like

```
SELECT datoms00.e AS `?x`,
       datoms01.v AS `?age`,
       datoms02.v AS `?height`
FROM datoms datoms00, datoms datoms01, datoms datoms02
WHERE datoms00.e = datoms01.e
  AND datoms00.e = datoms02.e
  AND …         -- Attribute checks go here.
```

If you add a `not` in there, and do what you've done here:

```
[:find ?x ?age ?height
 :where
 [?x :foo/name "John"]
 [?x :foo/age ?age]
 [?x :foo/height ?height]
 (not-join [?x] [?x :foo/degree :bsc])]
```

you'll get SQL like:

```
SELECT datoms00.e AS `?x`,
       datoms01.v AS `?age`,
       datoms02.v AS `?height`
FROM datoms datoms00, datoms datoms01, datoms datoms02
WHERE datoms00.e = datoms01.e
  AND datoms00.e = datoms02.e
  AND …         -- Attribute checks go here.
  AND NOT EXISTS (
    SELECT 1 FROM datoms datoms03
    WHERE datoms03.a = 99         -- :foo/degree
      AND datoms03.v = $bsc
      AND datoms03.e = datoms00.e
      AND datoms03.e = datoms01.e
      AND datoms03.e = datoms02.e
  )
```

Notice that because you've copied the full set of column bindings into the inner query, you'll get the full keyed expansion of external columns. That's redundant.

I believe a sufficient fix is to only use here the first column binding, yielding:

```
SELECT datoms00.e AS `?x`,
       datoms01.v AS `?age`,
       datoms02.v AS `?height`
FROM datoms datoms00, datoms datoms01, datoms datoms02
WHERE datoms00.e = datoms01.e
  AND datoms00.e = datoms02.e
  AND …         -- Attribute checks go here.
  AND NOT EXISTS (
    SELECT 1 FROM datoms datoms03
    WHERE datoms03.a = 99         -- :foo/degree
      AND datoms03.v = $bsc
      AND datoms03.e = datoms00.e
  )
```

What do you think?